src/Microsoft.OData.Core/ODataBatchReader.cs

```diff
@@ -204,440 +233,82 @@ void IODataBatchOperationListener.BatchOperationContentStreamDisposed()
         }
 
         /// <summary>
-        /// Returns the next state of the batch reader after an end boundary has been found.
+        /// Increases the size of the current batch message; throws if the allowed limit is exceeded.
         /// </summary>
-        /// <returns>The next state of the batch reader.</returns>
-        private ODataBatchReaderState GetEndBoundaryState()
+        protected void IncreaseBatchSize()
         {
-            switch (this.batchReaderState)
+            this.currentBatchSize++;

```

Comment:
>currentBatchSize [](start = 17, length = 16)

Seems like the validation should be done before the variable is incremented and left in an invalid state.

Reply:
This is the original way of how sizes are incremented in original ODataBatchReader/Writer ([https://github.com/OData/odata.net/blob/master/src/Microsoft.OData.Core/ODataBatchReader.cs#L763](https://github.com/OData/odata.net/blob/master/src/Microsoft.OData.Core/ODataBatchReader.cs#L763)).
When the batch size exceeds the max allowed, the processing of the batch has to be aborted.
Good observation though. There shouldn't be data-racing issue, therefore I guess we can change it if you have strong opinion on this.

---
In reply to: [140842889](https://github.com/OData/odata.net/pull/960#discussion_r140842889) [](ancestors = 140842889)